File: tools/balance_upgrades.py
#!/usr/bin/env python3
"""
Narzędzie do analizy balansu upgrade'ów vs capacity serwerów.
Używa tych samych formuł co gra (gameLogic.h / State.h).

Uruchomienie:
    python3 tools/balance_upgrades.py
"""

# === PARAMETRY GRY (z State.h / gameLogic.h) ===
MaxRacks = 3
RackSize = 4
MaxServerLevel = 20
serverLevelCapacityScale = 19

max_slots = MaxRacks * RackSize

SERVER_RATE = 1.3

def server_bonus(level):
    """getServerUpgradeBonus z gameLogic.h: compound 1.4x per level"""
    if level == 0:
        return 0
    bonus = serverLevelCapacityScale
    for i in range(1, level):
        bonus = int(bonus * SERVER_RATE)
    return bonus

max_capacity = max_slots * server_bonus(MaxServerLevel)

# Formuła bonusu z gameLogic.h: getOfficeUpgradeBonus()
# bonus = startingBonus; for each level: bonus += bonus * COMPOUND_RATE
COMPOUND_RATE = 0.5

upgrades = [
    {"name": "User messaging",       "price": 80,          "bonus": 10,   "max": 10},
    {"name": "Push notifications",   "price": 1000,        "bonus": 22,   "max": 10},
    {"name": "Load balancing",       "price": 3500,        "bonus": 45,   "max": 10},
    {"name": "Request routing",      "price": 12000,       "bonus": 60,   "max": 10},
    {"name": "Async processing",     "price": 40000,       "bonus": 85,   "max": 10},
    {"name": "AI core upgrade",      "price": 200000,      "bonus": 120,  "max": 5},
    {"name": "AI Context awareness", "price": 400000,      "bonus": 200,  "max": 5},
    {"name": "AI Smart responses",   "price": 4000000,     "bonus": 400,  "max": 5},
    {"name": "Predictive AI",        "price": 20000000,    "bonus": 800,  "max": 5},
    {"name": "Self evolving AI",     "price": 800000000,   "bonus": 1000, "max": 1},
]

# === FORMUŁY ===

def bonus_compound(sb, have, rate=COMPOUND_RATE):
    """Obecna formuła z gry (getOfficeUpgradeBonus)"""
    if have == 0:
        return 0
    b = sb
    for i in range(have):
        b += int(b * rate)
    return b


# === ANALIZA ===

def analyze(label, upgrade_set, bonus_func, show_steps=False):
    print(f"\n{'=' * 100}")
    print(f"  {label}")
    print(f"{'=' * 100}")

    # Podsumowanie per upgrade
    print(f"  {'#':>2} | {'Upgrade':25s} | {'Bonus':>6} | {'Max':>3} | {'Contrib':>10} | {'Cumul':>10} | {'% cap':>8}")
    cumul = 1
    for i, u in enumerate(upgrade_set):
        c = bonus_func(u["bonus"], u["max"])
        cumul += c
        pct = cumul / max_capacity * 100
        bar_len = min(int(pct / 5), 20)
        bar = "#" * bar_len + "." * (20 - bar_len)
        overflow = " OVER" if cumul > max_capacity else ""
        print(f"  {i:>2} | {u['name']:25s} | {u['bonus']:>6} | {u['max']:>3} | {c:>10,} | {cumul:>10,} | [{bar}] {pct:>5.1f}%{overflow}")
    print(f"  TOTAL: {cumul:,} / {max_capacity:,} = {cumul/max_capacity*100:.1f}%")

    # Krok po kroku
    if show_steps:
        print(f"\n  Inbound po kazdym zakupie:")
        cumul = 1
        current = [0] * len(upgrade_set)
        pnum = 0
        for i, u in enumerate(upgrade_set):
            for lvl in range(1, u["max"] + 1):
                pnum += 1
                current[i] = lvl
                cumul = 1 + sum(
                    bonus_func(upgrade_set[j]["bonus"], current[j])
                    for j in range(len(upgrade_set))
                )
                pct = cumul / max_capacity * 100
                bar_len = min(int(pct / 5), 20)
                bar = "#" * bar_len + "." * (20 - bar_len)
                overflow = " OVER" if cumul > max_capacity else ""
                print(f"    #{pnum:>3} {u['name']:25s} lvl {lvl:>2} => inbound={cumul:>6,} [{bar}] {pct:>5.1f}%{overflow}")

    return cumul


def find_overflow_point(upgrade_set, bonus_func):
    current = [0] * len(upgrade_set)
    p = 0
    for i, u in enumerate(upgrade_set):
        for lvl in range(1, u["max"] + 1):
            p += 1
            current[i] = lvl
            ti = 1 + sum(bonus_func(upgrade_set[j]["bonus"], current[j]) for j in range(len(upgrade_set)))
            if ti > max_capacity:
                print(f"\n  OVERFLOW przy: {u['name']} lvl {lvl} (zakup #{p})")
                print(f"  inbound={ti:,} > capacity={max_capacity:,}")
                return
    print(f"\n  Brak overflow - max inbound miesci sie w capacity")


# === MAIN ===

if __name__ == "__main__":
    print(f"MAX CAPACITY = {max_slots} slots x server_bonus({MaxServerLevel})={server_bonus(MaxServerLevel)} = {max_capacity}")

    print("\n" + "=" * 100)
    print("  CAPACITY na roznych etapach gry")
    print("=" * 100)
    milestones = [
        ("Start (1 srv lvl 1)",              1 * server_bonus(1)),
        ("1 rack filled (4 srv lvl 1)",       4 * server_bonus(1)),
        ("1 rack maxed (4 srv lvl 20)",       4 * server_bonus(MaxServerLevel)),
        ("3 racks filled lvl 5",             12 * server_bonus(5)),
        ("3 racks filled lvl 10",            12 * server_bonus(10)),
        ("3 racks filled lvl 15",            12 * server_bonus(15)),
        ("ALL MAXED",                         max_capacity),
    ]
    for label, cap in milestones:
        print(f"  {label:35s} => capacity = {cap:>6,}")

    # --- SYMULACJA LEVELI SERWERA ---
    print("\n" + "=" * 100)
    print("  SERVER LEVELS (1 serwer, lvl 1-20)")
    print("=" * 100)
    serverPrice = 1000
    print(f"  {'Lvl':>3} | {'Capacity':>10} | {'Gain':>6} | {'Upgrade cost':>14} | {'Cumul cost':>14} | {'Cost/cap':>10}")
    cumul_cost = 0
    prev_cap = 0
    for lvl in range(1, MaxServerLevel + 1):
        cap = server_bonus(lvl)
        gain = cap - prev_cap
        cost = int(serverPrice * (1.4 ** lvl))
        cumul_cost += cost
        cost_per_cap = cumul_cost / cap if cap > 0 else 0
        print(f"  {lvl:>3} | {cap:>10,} | +{gain:>4} | {cost:>14,} | {cumul_cost:>14,} | {cost_per_cap:>10.1f}")
        prev_cap = cap

    # --- OBECNE WARTOSCI ---
    analyze(
        "OBECNE WARTOSCI (compound 50%)",
        upgrades,
        lambda sb, h: bonus_compound(sb, h, 0.5),
        show_steps=True,
    )
    find_overflow_point(upgrades, lambda sb, h: bonus_compound(sb, h, 0.5))

File: tools/balance_upgrades.cpp
/*
 * Narzędzie do analizy balansu upgrade'ów vs capacity serwerów.
 * Includuje bezpośrednio gameLogic.h / State.h — jedno źródło prawdy.
 *
 * Kompilacja i uruchomienie:
 *   cd tools && make
 *   ./balance_upgrades
 */

#include <cstdio>
#include <cstdint>
#include <cmath>
#include <cstring>

// PROGMEM musi być zdefiniowany przed State.h (który go używa a nie includuje Arduboy2.h)
#define PROGMEM

// === INCLUDE PLIKÓW GRY (jedno źródło prawdy) ===
#include "../Sketch/State.h"
#include "../Sketch/gameLogic.h"

// === HELPER: formatowanie liczb z separatorami tysięcy ===
void printNum(char* buf, size_t sz, int64_t n) {
    if (n < 0) { buf[0] = '-'; printNum(buf+1, sz-1, -n); return; }
    if (n < 1000) { snprintf(buf, sz, "%lld", (long long)n); return; }
    char tmp[32];
    printNum(tmp, sizeof(tmp), n / 1000);
    snprintf(buf, sz, "%s,%03lld", tmp, (long long)(n % 1000));
}

const char* fmt(int64_t n) {
    static char bufs[4][32];
    static int idx = 0;
    idx = (idx + 1) % 4;
    printNum(bufs[idx], sizeof(bufs[idx]), n);
    return bufs[idx];
}

void printBar(double pct) {
    int barLen = (int)(pct / 5);
    if (barLen > 20) barLen = 20;
    printf("[");
    for (int i = 0; i < 20; i++) printf("%c", i < barLen ? '#' : '.');
    printf("]");
}

void printSep() {
    for (int i = 0; i < 100; i++) printf("=");
    printf("\n");
}

// === MAIN ===

int main() {
    const int maxSlots = MaxRacks * RackSize;
    const uint32_t maxCap = maxSlots * getServerUpgradeBonus(MaxServerLevel);

    // --- PARAMETRY ---
    printf("PARAMETRY Z PLIKÓW GRY:\n");
    printf("  MaxRacks=%d, RackSize=%d, MaxServerLevel=%d\n", MaxRacks, RackSize, MaxServerLevel);
    printf("  serverPrice=%d, serverLevelCapacityScale=%d\n", serverPrice, serverLevelCapacityScale);
    printf("  Upgrades: %d\n", MaxUpgrades);
    for (int i = 0; i < MaxUpgrades; i++) {
        printf("    %d: %-25s price=%10s  bonus=%5d  max=%d\n",
            i, (const char*)upgradesNames[i],
            fmt(upgrades[i].startingPrice), upgrades[i].startingBonus, upgrades[i].max);
    }

    printf("\nMAX CAPACITY = %d slots x server_bonus(%d)=%s = %s\n",
        maxSlots, MaxServerLevel, fmt(getServerUpgradeBonus(MaxServerLevel)), fmt(maxCap));

    // --- CAPACITY NA RÓŻNYCH ETAPACH ---
    printf("\n"); printSep();
    printf("  CAPACITY na roznych etapach gry\n");
    printSep();

    struct { const char* label; uint32_t cap; } milestones[] = {
        {"Start (1 srv lvl 1)",          1 * getServerUpgradeBonus(1)},
        {"1 rack filled (4 srv lvl 1)",  4 * getServerUpgradeBonus(1)},
        {"1 rack maxed (4 srv lvl 20)",  4 * getServerUpgradeBonus(MaxServerLevel)},
        {"3 racks filled lvl 5",        12 * getServerUpgradeBonus(5)},
        {"3 racks filled lvl 10",       12 * getServerUpgradeBonus(10)},
        {"3 racks filled lvl 15",       12 * getServerUpgradeBonus(15)},
        {"ALL MAXED",                    maxCap},
    };
    for (auto& m : milestones) {
        printf("  %-35s => capacity = %6s\n", m.label, fmt(m.cap));
    }

    // --- SERVER LEVELS ---
    printf("\n"); printSep();
    printf("  SERVER LEVELS (1 serwer, lvl 1-20)\n");
    printSep();
    printf("  %3s | %10s | %6s | %14s | %14s | %10s\n",
        "Lvl", "Capacity", "Gain", "Upgrade cost", "Cumul cost", "Cost/cap");
    uint32_t cumulCost = 0;
    uint32_t prevCap = 0;
    for (int lvl = 1; lvl <= MaxServerLevel; lvl++) {
        uint32_t cap = getServerUpgradeBonus(lvl);
        uint32_t gain = cap - prevCap;
        uint32_t cost = getServerUpgradeCost(lvl);
        cumulCost += cost;
        double costPerCap = cap > 0 ? (double)cumulCost / cap : 0;
        printf("  %3d | %10s | +%4s | %14s | %14s | %10.1f\n",
            lvl, fmt(cap), fmt(gain), fmt(cost), fmt(cumulCost), costPerCap);
        prevCap = cap;
    }

    // --- KOSZTY UPGRADE'ÓW ---
    printf("\n"); printSep();
    printf("  KOSZTY UPGRADE'OW (price * 1.5^level)\n");
    printSep();
    for (int i = 0; i < MaxUpgrades; i++) {
        Upgrade& u = upgrades[i];
        printf("\n  %s (base price: %s, max lvl: %d)\n",
            (const char*)upgradesNames[i], fmt(u.startingPrice), u.max);
        printf("  %3s | %14s | %14s | %10s | %10s\n",
            "Lvl", "Koszt lvl", "Koszt cumul", "Bonus", "Cost/bonus");
        uint32_t cumCost = 0;
        for (int lvl = 1; lvl <= u.max; lvl++) {
            Upgrade tmp = {u.startingPrice, u.startingBonus, (uint8_t)lvl, u.max};
            uint32_t cost = getOfficeUpgradeCost(Upgrade{u.startingPrice, u.startingBonus, (uint8_t)lvl, u.max});
            cumCost += cost;
            uint32_t bonus = getOfficeUpgradeBonus(tmp);
            double cpb = bonus > 0 ? (double)cumCost / bonus : 0;
            printf("  %3d | %14s | %14s | %10s | %10.1f\n",
                lvl, fmt(cost), fmt(cumCost), fmt(bonus), cpb);
        }
    }

    // Podsumowanie kosztów
    printf("\n  %2s | %-25s | %14s | %14s | %14s\n",
        "#", "Upgrade", "Koszt 1 lvl", "Koszt max lvl", "Koszt total");
    uint64_t grandTotal = 0;
    for (int i = 0; i < MaxUpgrades; i++) {
        Upgrade& u = upgrades[i];
        uint32_t costFirst = getOfficeUpgradeCost(Upgrade{u.startingPrice, u.startingBonus, 1, u.max});
        uint32_t costLast = getOfficeUpgradeCost(Upgrade{u.startingPrice, u.startingBonus, u.max, u.max});
        uint64_t total = 0;
        for (int lvl = 1; lvl <= u.max; lvl++) {
            total += getOfficeUpgradeCost(Upgrade{u.startingPrice, u.startingBonus, (uint8_t)lvl, u.max});
        }
        grandTotal += total;
        printf("  %2d | %-25s | %14s | %14s | %14s\n",
            i, (const char*)upgradesNames[i], fmt(costFirst), fmt(costLast), fmt((int64_t)total));
    }
    printf("  %2s | %-25s | %14s | %14s | %14s\n",
        "", "GRAND TOTAL", "", "", fmt((int64_t)grandTotal));

    // --- ANALIZA: OBECNE WARTOŚCI ---
    printf("\n"); printSep();
    printf("  OBECNE WARTOSCI (compound 50%%)\n");
    printSep();

    // Podsumowanie per upgrade
    printf("  %2s | %-25s | %6s | %3s | %10s | %10s | %8s\n",
        "#", "Upgrade", "Bonus", "Max", "Contrib", "Cumul", "% cap");
    uint32_t cumul = 1;
    for (int i = 0; i < MaxUpgrades; i++) {
        Upgrade tmp = {upgrades[i].startingPrice, upgrades[i].startingBonus, upgrades[i].max, upgrades[i].max};
        uint32_t c = getOfficeUpgradeBonus(tmp);
        cumul += c;
        double pct = (double)cumul / maxCap * 100;
        printf("  %2d | %-25s | %6d | %3d | %10s | %10s | ",
            i, (const char*)upgradesNames[i], upgrades[i].startingBonus, upgrades[i].max,
            fmt(c), fmt(cumul));
        printBar(pct);
        printf(" %5.1f%%%s\n", pct, cumul > maxCap ? " OVER" : "");
    }
    printf("  TOTAL: %s / %s = %.1f%%\n", fmt(cumul), fmt(maxCap), (double)cumul/maxCap*100);

    // Krok po kroku
    printf("\n  Inbound po kazdym zakupie:\n");
    cumul = 1;
    uint8_t current[MaxUpgrades] = {};
    int pnum = 0;
    for (int i = 0; i < MaxUpgrades; i++) {
        for (int lvl = 1; lvl <= upgrades[i].max; lvl++) {
            pnum++;
            current[i] = lvl;
            uint32_t totalInbound = 1;
            for (int j = 0; j < MaxUpgrades; j++) {
                Upgrade tmp = {upgrades[j].startingPrice, upgrades[j].startingBonus, current[j], upgrades[j].max};
                totalInbound += getOfficeUpgradeBonus(tmp);
            }
            double pct = (double)totalInbound / maxCap * 100;
            printf("    #%3d %-25s lvl %2d => inbound=%6s ",
                pnum, (const char*)upgradesNames[i], lvl, fmt(totalInbound));
            printBar(pct);
            printf(" %5.1f%%%s\n", pct, totalInbound > maxCap ? " OVER" : "");
        }
    }

    // Overflow point
    memset(current, 0, sizeof(current));
    pnum = 0;
    bool foundOverflow = false;
    for (int i = 0; i < MaxUpgrades && !foundOverflow; i++) {
        for (int lvl = 1; lvl <= upgrades[i].max; lvl++) {
            pnum++;
            current[i] = lvl;
            uint32_t ti = 1;
            for (int j = 0; j < MaxUpgrades; j++) {
                Upgrade tmp = {upgrades[j].startingPrice, upgrades[j].startingBonus, current[j], upgrades[j].max};
                ti += getOfficeUpgradeBonus(tmp);
            }
            if (ti > maxCap) {
                printf("\n  OVERFLOW przy: %s lvl %d (zakup #%d)\n",
                    (const char*)upgradesNames[i], lvl, pnum);
                printf("  inbound=%s > capacity=%s\n", fmt(ti), fmt(maxCap));
                foundOverflow = true;
                break;
            }
        }
    }
    if (!foundOverflow) {
        printf("\n  Brak overflow - max inbound miesci sie w capacity\n");
    }

    return 0;
}
